### PR TITLE
Smaller syncer cache

### PR DESCRIPTION
**Pull request by robbrockbank**
Convert the watcher-synced cache to be a map[string]string rather than storing a struct containing the key and revision.

The cache now uses:

`map[<defaultpath>]<revision>`   i.e.  `map[string]string`

I'm not sure if this will help or not since the default paths can be a little verbose.  We could certainly improve things further by using a contracted path (i.e. one that contains the minimum info) - the key thing is we can convert back again from the path to the Key.

### CODE REVIEW

File: lib/backend/watchersyncer/watchercache.go [hunk 1]
```diff
@@ -201,11 +192,12 @@ func (wc *watcherCache) finishResync() {
 	if numOldResources > 0 {
 		wc.logger.WithField("Num", numOldResources).Debug("Sending resync deletes")
 		updates := make([]api.Update, 0, len(wc.oldResources))
-		for _, r := range wc.oldResources {
+		for p, _ := range wc.oldResources {
+			key := model.KeyFromDefaultPath(p)
```
Comment (fasaxc) [inline lib/backend/watchersyncer/watchercache.go]: Worth asserting that key is not nil?  We should always be able to parse the keys we have here

File: lib/backend/model/keys.go [hunk 2]
```diff
@@ -185,65 +191,16 @@ func IsListOptionsLastSegmentPrefix(listOptions ListInterface) bool {
...
-			WorkloadID:     unescapeName(m[3]),
-			EndpointID:     unescapeName(m[4]),
-		}
-	} else if m := matchHostEndpoint.FindStringSubmatch(path); m != nil {
-		log.Debugf("Path is a host endpoint: %v", path)
-		return HostEndpointKey{
-			Hostname:   m[1],
-			EndpointID: unescapeName(m[2]),
+	log.WithField("path", path).Info("KeyFromDefaultPath")
+	for _, lo := range listTypes {
```
[the remark this answers is not in the document]
Comment (robbrockbank) [reply lib/backend/model/keys.go]: Yeah I was worrying about that - in fact I was thinking it'd be better to just register all of the resources in a single init() block which means we can tune the overall order for these checks.

I had considered registering a slice of segments and doing this matching using a trie.

**Comment by caseydavenport** [comment]
@robbrockbank @fasaxc

Is this PR required for Calico v3.0?

**Comment by robbrockbank** [comment]
Closing - may resurrect in future, but as a different PR.

**Pull request closed without merging**